src/block_cache/block_tier.c

```diff
@@ -121,30 +123,44 @@ __wt_blkcache_get_handle(WT_SESSION_IMPL *session, WT_BM *bm, uint32_t objectid,
             __wt_readunlock(session, &bm->handle_array_lock);
             return (0);
         }
-
-    /* We need to add a new handle the block handle array. Upgrade to a write lock. */
     __wt_readunlock(session, &bm->handle_array_lock);
+
+    /* Open a handle for the object. */
+    WT_RET(__wt_blkcache_tiered_open(session, NULL, objectid, &new_handle));
+
+    /* We need a write lock to add a new entry to the handle array. */
     __wt_writelock(session, &bm->handle_array_lock);
 
-    /* Check to make sure the object wasn't cached while we locked. */
-    for (i = 0; i < bm->handle_array_next; ++i)
+    /* Check to see if the object was added while we opened it. */
+    for (i = 0, found = false; i < bm->handle_array_next && !found; ++i)
         if (bm->handle_array[i]->objectid == objectid) {
             *blockp = bm->handle_array[i];
-            __wt_writeunlock(session, &bm->handle_array_lock);
-            return (0);
+            found = true;
         }
```

Comment:
I think some refactoring of the loop into a different function would help.  Assume lines 134-137 become a little function like `void __blkcache_find_object_block(WT_BM *bm, WT_BLOCK **blockp)`.  Then 119-140 becomes:
```
  __wt_readlock(session, &bm->handle_array_lock);
  __blockcache_find_object_block(bm, blockp);
  __wt_readunlock(session, &bm->handle_array_lock);
  if (*blockp != NULL)
    return (0);

  /* Open a handle for the object. */
  WT_RET(__wt_blkcache_tiered_open(session, NULL, objectid, &new_handle));

  /* We need a write lock to add a new entry to the handle array. */
  __wt_writelock(session, &bm->handle_array_lock);
  __blockcache_find_object_block(bm, blockp);
  ...
```
We don't need `found` as `*block != NULL` serves the purpose.  I find it easier to reason about (esp the locks).

Reply:
I had a similar thought -- that we should refactor the manipulation of the handle array into its own set of functions. In addition to making the code simpler, as you point out here, it would also make it simpler if/when in the future we decide we want to use a different data structure for tracking open handles. 

I was thinking of doing that refactoring in a future ticket so we could also think about the other places where we iterate through the handle array (during checkpoint and close). But I like this change on its own merits. So I'll go ahead with it now.

Thanks!